### ISSUE TITLE
Is it possible to specify code to run when the server is stopped?

### ISSUE THREAD

**Issue by atheriel** [rstudio/httpuv#179]
I have a server where I'd like to persist some state to disk on shutdown, allowing the program to pick up where it left off. As a contrived example, imagine a server that reads a counter from a file and increments that counter when it receives a request:

```r
count_server <- function(fname) {
  counter <- if (file.exists(fname)) {
    unserialize(fname)
  } else {
    0
  }
  httpuv::startServer("0.0.0.0", 8080, list(call = function(req) {
    counter <<- counter + 1
    list(
      status = 200L,
      headers = c("Content-Type" = "text/plain; charset=utf-8"),
      body = paste0("Counter: ", counter)
    )
  }))
}

handle <- count_server(tempfile())

# Some time later...
httpuv::stopServer(handle)
```

At the moment, I don't see a way of ensuring that this counter can be written out to disc when `stopServer()` is called. Is it possible that `startServer()` could take a callback to be run on shutdown? Or is there some other way of handling these kinds of issues?

**Comment by wch** [comment]
There isn't currently a way to do this, but because of some recent changes, adding this functionality should be too difficult.

With the current development version of httpuv, `startServer()` now returns an `WebServer` R6 object instead of an arbitrary string handle. The object has a method, `$stop()`, which stops the app (`stopServer(app)` now is simply a function that calls `app$stop()`).

The [constructor](https://github.com/rstudio/httpuv/blob/a830a0d/R/server.R#L164-L187) for `WebServer` could be modified so that `app` can also contain callback functions for various events, and the same could be done for `PipeServer`. For example, I think that these make sense:

* `onStart`: called just before the server starts. (Not sure this one is necessary, though.)
* `onStarted`: called just after the server starts.
* `onStop`: called just before the server stops.
* `onStopped`: called just after the server stops.

There could also be an option for `$stop()` (and therefore the `onStopped` callback) to be called by a finalizer for the `WebServer`, if the server is still running at the time. This would only be invoked in the case where the R process exits while the server is still running, because httpuv keeps references to all of the running `WebServer`/`PipeServer` objects in a list -- the object representing a running server won't be GC'd. Finalizers are a little unsafe to call, which is why this behavior should be optional.

**Comment by jcheng5** [comment]
I'd consider writing to disk every time. That way even if the process is terminated or the server shuts down, you don't lose much.

**Comment by atheriel** [comment]
@jcheng5 Yeah, I realize that's a simple resolution in this case. I'm happy to close the issue, unless you would like to leave it open for discussion of @wch's suggestions.
